File: js/src/resources/speech-to-text.ts
import type { HttpClient, PollingOptions, RequestOptions } from '@runapi.ai/core';
import { compactParams } from '@runapi.ai/core';
import { pollUntilComplete } from '@runapi.ai/core/internal';
import type { CompletedSpeechToTextResponse, TaskCreateResponse, SpeechToTextParams, SpeechToTextResponse } from '../types';

const ENDPOINT = '/api/v1/elevenlabs/speech_to_text';

export class SpeechToText {
  constructor(private readonly http: HttpClient) {}

  async run(params: SpeechToTextParams, options?: RequestOptions & PollingOptions): Promise<CompletedSpeechToTextResponse> {
    const { id } = await this.create(params, options);
    const response = await pollUntilComplete<SpeechToTextResponse>(() => this.get(id, options), {
      maxWaitMs: options?.maxWaitMs,
      pollIntervalMs: options?.pollIntervalMs,
    });
    return response as CompletedSpeechToTextResponse;
  }

  async create(params: SpeechToTextParams, options?: RequestOptions): Promise<TaskCreateResponse> {
    return this.http.request<TaskCreateResponse>('POST', ENDPOINT, {
      body: compactParams(params),
      ...options,
    });
  }

  async get(id: string, options?: RequestOptions): Promise<SpeechToTextResponse> {
    return this.http.request<SpeechToTextResponse>('GET', `${ENDPOINT}/${id}`, options ?? {});
  }
}

File: ruby/lib/runapi/elevenlabs/resources/speech_to_text.rb
# frozen_string_literal: true

module RunApi
  module Elevenlabs
    module Resources
      class SpeechToText
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/elevenlabs/speech_to_text"
        RESPONSE_CLASS = Types::SpeechToTextResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedSpeechToTextResponse

        def initialize(http)
          @http = http
        end

        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        def create(**params)
          params = compact_params(params)
          raise Core::ValidationError, "audio_url is required" unless param(params, :audio_url)
          request(:post, ENDPOINT, body: params)
        end

        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end
      end
    end
  end
end
